src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -902,6 +903,30 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
             return false;
         }
 
+        public static bool IsFunctionPointerTypeArgumentContext(
+            this SyntaxTree syntaxTree,
+            int position,
+            SyntaxToken tokenOnLeftOfPosition,
+            CancellationToken cancellationToken)
+        {
+            var token = tokenOnLeftOfPosition;
+            token = token.GetPreviousTokenIfTouchingWord(position);
+
+            switch (token.Kind())
+            {
+                case SyntaxKind.LessThanToken:
+                case SyntaxKind.CommaToken:
+                    return token.Parent is FunctionPointerTypeSyntax;
+            }
+
+            if (token.GetAncestor<FunctionPointerTypeSyntax>() == null)
+            {
+                return false;
+            }
+
+            return syntaxTree.IsInPartiallyWrittenFunctionPointer(position, cancellationToken);
```

Comment:
the partially-written stuff is necessary for normal generics, which are ambiguous with expressions.  i.e. `List<a` could be an expr or a generic type.  for func-ptrs, we never have that ambiguity.  You have `delegate*` so we always know its' a funcptr.

Reply:
Simplified.